Add (iid) truncated normal and also add grad to normal

chaozg opened the pull request:
fixed #321 to add `TruncatedNormal`
fixed #548 to add `gradient` to `Normal`
fixed #539 to add a demo on `TruncatedNormal`

## Description

- With this PR, we can draw samples from truncated iid normal distributions, i.e., truncated version of `cuqi.distribution.normal` and also use them as priors. 
- Gibbs with a gradient sampler works also fine on it.
- Plots from demo38 can be seen at #549 .

## Test with Gibbs
```python
import cuqi
import numpy as np
np.random.seed(0)
d = cuqi.distribution.Uniform(np.array([0]),np.array([1]))
A = cuqi.model.LinearModel(lambda x: x, adjoint=lambda x: x, range_geometry=1, domain_geometry=1)
x = cuqi.distribution.TruncatedNormal(A(d), std=np.array([1]), low=np.array([-1]), high=np.array([np.Inf]))

sampling_strategy = {
    "d" : cuqi.experimental.mcmc.MALA(initial_point=np.array([0.5]), scale=0.1),
    "x" : cuqi.experimental.mcmc.MALA(initial_point=np.array([1.0]), scale=0.3)
}
joint = cuqi.distribution.JointDistribution(d, x)
sampler = cuqi.experimental.mcmc.HybridGibbs(joint, sampling_strategy=sampling_strategy)
sampler.sample(10000)
samples = sampler.get_samples()
samples["x"].plot_trace()
samples["d"].plot_trace()
```

(Note: here `x` is truncated from -1, just as expected)
![image](https://github.com/user-attachments/assets/0639db0f-8fbe-4f13-b5dd-d91ca7b59b80)
![image](https://github.com/user-attachments/assets/f9427fbc-3300-4037-98d2-d871e9766818)

amal-ghamdi commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,48 @@
+
+class TruncatedNormal(Distribution):
+    """
+    Truncated Normal probability distribution. Generates instance of cuqi.distribution.TruncatedNormal.
```
We can possibly add a bit more description about the distribution: e.g. 
```suggestion
    Truncated Normal probability distribution. Generates instance of cuqi.distribution.TruncatedNormal. It allows the user to specify upper and lower bounds on random variables represented by a Normal distribution. This distribution is suitable for a small dimension setup (e.g. `dim`=3 or 4). Using TruncatedNormal Distribution with a larger dimension can lead to a high rejection rate when used within MCMC samplers.
```

amal-ghamdi commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,48 @@
+    ------------
+    mean: mean of distribution
+    std: standard deviation
+    a: lower bound of the distribution
+    b: upper bound of the distribution
```
I am not sure what is the best name for the upper and lower bounds, I suggest, however, making them consistent with the uniform distribution. In the uniform distribution, we have 
```
        low : float or array_like of floats
            Lower bound(s) of the uniform distribution.
        high : float or array_like of floats 
            Upper bound(s) of the uniform distribution.
```
 It could be we also use `low` and `high` here, or we change the names in both distributions to `lower` and `upper` or something, what do you think?

nabriis replied:
Sounds good to me.

amal-ghamdi replied:
👍

amal-ghamdi commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,48 @@
+    The variables of this distribution are iid.
+
+    
+    Parameters
```
I suggest adding types to the parameters, e.g. 
```
    mean: scalar or 1d-array
        Mean of the distribution
```

chaozg replied:
Done as in uniform:
```python
    mean : float or array_like of floats
        mean of distribution
    std : float or array_like of floats
        standard deviation
    low : float or array_like of floats
        lower bound of the distribution
    high : float or array_like of floats
        upper bound of the distribution
```

amal-ghamdi replied:
👍

amal-ghamdi commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,48 @@
+        if np.any(x < self.a) or np.any(x > self.b):
+            return -np.Inf
+        else:
+             return np.sum(-np.log(self.std*np.sqrt(2*np.pi))-0.5*((x-self.mean)/self.std)**2)
```
One trick we did in lognormal distribution is that we sat up an underlying Gaussian and used its logpdf and gradient. This gives more robustness in general because we relay on a single implementation that is heavily tested. Another advantage I believe is that we get the chain rule for the gradient for free in this case (e.g. when lognormal is used for likelihood for example). You can consider using this trick here as well.

chaozg replied:
That's an excellent suggestion and following this strategy it will become way easier to adapt a truncated version of `cuqi.distribution.Gaussian` (instead of the iid `cuqi.distribution.Normal`)

amal-ghamdi replied:
👍

amal-ghamdi commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,48 @@
+        """
+        Generates random samples from the distribution.
+        """
+        raise NotImplementedError(f"sample is not implemented for {self.__class__.__name__}.")
```
This is a general comment, it could be nice to add some tests like the ones we have in [test_distribution.py](https://github.com/CUQI-DTU/CUQIpy/blob/main/tests/test_distribution.py). This will make sure that if someone update this code in the future, the updates still pass the test. See for example the specific tests: `test_lognormal_logpdf`, `test_gradient_lognormal_as_prior`, and `test_gradient_lognormal_as_likelihood`.

nabriis replied:
Agreed

chaozg replied:
Added `test_TruncatedNormal_logpdf` and `test_TruncatedNormal_gradient`

amal-ghamdi replied:
👍

amal-ghamdi reviewed (changes requested):
Thank you @chaozg for adding this very useful distribution! I added some comments, feel free to address them as you see fit.

nabriis commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,48 @@
+
+class TruncatedNormal(Distribution):
+    """
+    Truncated Normal probability distribution. Generates instance of cuqi.distribution.TruncatedNormal. It allows the user to specify upper and lower bounds on random variables represented by a Normal distribution. This distribution is suitable for a small dimension setup (e.g. `dim`=3 or 4). Using TruncatedNormal Distribution with a larger dimension can lead to a high rejection rate when used wi ...
```
Could you break this line into smaller pieces?

chaozg replied:
Done as suggested

nabriis commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,48 @@
+        else:
+            return -(x-self.mean)/(self.std**2)
+
+    def _sample(self,N=1, rng=None):
```
```suggestion
    def _sample(self, N=1, rng=None):
```

nabriis commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,48 @@
+
+    def _sample(self,N=1, rng=None):
+        """
+        Generates random samples from the distribution.
```
Perhaps for the direct sampling of the distribution we could use a simple "Rejection" sampling? If asked for "N" samples, we sample a Gaussian, say "N" times, remove out of bounds, sample another "N" times, remote out of bounds. If we have now "N" or more sample we return the first N, else we repeat until we get "N" samples. Could also be used to compare with the samplers you showed in the code.

chaozg replied:
I just added`_sample` with a slightly different strategy: I generate samples one by one, check if it is within the bounds, until we get N samples. Do you think strategy is OK?

amal-ghamdi replied:
just a comment here, would it be possible just to pass the rng to `sample = self._normal.sample(rng=rng)`?

chaozg replied:
I think you're right @amal-ghamdi; I just updated the code as you suggested. Thanks!

nabriis commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,48 @@
+from scipy.special import erf
+from cuqi.distribution import Distribution
+
+class TruncatedNormal(Distribution):
```
You made this really nice showcases in the PR @chaozg. I suggest you add one or two of them as "Examples" in the docstring!

chaozg replied:
Added a example in the docstring as suggested

nabriis reviewed (changes requested):
Really nice @chaozg. I agree with Amals suggestions and I had a few of my own.

chaozg commented:
Thanks @amal-ghamdi and @nabriis for your review. I think I have addressed your comments so I'm requesting your further review here.

Note that the existing Normal lacks grad(), so I also add it in this PR.
The sample() lack a way to carry rng, and I'm not sure what's the best way to do this and would like to listen to your advice at stand-up : )

amal-ghamdi commented on cuqi/distribution/_normal.py:
```diff
@@ -36,6 +36,9 @@ def logpdf(self, x):
     def cdf(self, x):
         return np.prod(0.5*(1 + erf((x-self.mean)/(self.std*np.sqrt(2)))))
 
+    def gradient(self, x):
+        return -(x-self.mean)/(self.std**2)
+
```
This is a nice bonus that we have a gradient for normal now :). Just one suggestion to add two checks:
1- the geometry is not a geometry that needs a chain rule. 
2- The distribution is used as prior, not likelihood because we do not account for the chain rule here. 

For example, the gradient implementation of [cmrf](https://github.com/CUQI-DTU/CUQIpy/blob/main/cuqi/distribution/_cmrf.py) does these two checks



```python
    def _gradient(self, val, **kwargs):
        #Avoid complicated geometries that change the gradient.
        if not type(self.geometry) in _get_identity_geometries():
            raise NotImplementedError("Gradient not implemented for distribution {} with geometry {}".format(self,self.geometry))

        if not callable(self.location): # for prior
            diff = self._diff_op._matrix @ val
            return (-2*diff/(diff**2+self.scale**2)) @ self._diff_op._matrix
        else:
            warnings.warn('Gradient not implemented for {}'.format(type(self.location)))
```

amal-ghamdi commented on tests/test_distribution.py:
```diff
@@ -28,6 +28,73 @@ def test_Normal_sample_regression(mean,var,expected):
+        if np.all(point >= low) and np.all(point <= high):
+            assert x_trun.logpdf(point) == approx(x.logpdf(point))
+        else:
+            assert np.isinf(x_trun.logpdf(point))
```
a suggestion to use `np.isneginf` instead of `np.isinf` here, to check for negative infinity.

chaozg replied:
Done as suggested

amal-ghamdi reviewed (changes requested):
Many thanks @chaozg for the updates! it is nice that the distribution is based on Normal now and that you added the gradient for the Normal distribution. I have only few comments to consider as you see fit.

amal-ghamdi reviewed (approved):
LGTM!

nabriis commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,89 @@
+        self.high = high
+
+        # Init underlying normal distribution
+        self._normal = Normal(self.mean, self.std, is_symmetric=is_symmetric, **kwargs)
```
I think this will create issues with Gibbs sampling. We need to have getters and setters for mean and std, and a setter for _normal.

chaozg replied:
Done as suggested

chaozg replied:
Just added a MWE on this and corresponding plots at the bottom of my main comment

nabriis commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        Generates random samples from the distribution.
+        """
+        # FIXME: this implementation does not honor the rng
```
```suggestion

```

chaozg commented:
With @amal-ghamdi 's help today, `gradient` of TruncatedNormal finally works fine as a likelihood 😄 . And now I'm requesting further review from @nabriis .

chaozg commented:
Hi @amal-ghamdi since I just added a demo38 on the use of `TruncatedNormal`, I will request your further review again.

amal-ghamdi commented on cuqi/distribution/_normal.py:
```diff
@@ -42,10 +62,15 @@ def _gradient(self, val, *args, **kwargs):
+            dev = val - model.forward(*args, **kwargs)
+            if isinstance(dev, numbers.Number):
+                dev = np.array([dev])
+            return model.gradient(1.0/(np.array(self.std)) @ dev, *args, **kwargs)
```
I wonder if  "np.array" is still needed now that you have the std setter?

chaozg replied:
Yes, `dev` is already of `np.array` type. So I removed the line 68 and 69 as suggested.

amal-ghamdi commented on cuqi/distribution/_truncated_normal.py:
```diff
@@ -47,7 +48,47 @@ def __init__(self, mean=None, std=None, low=-np.Inf, high=np.Inf, is_symmetric=F
 
         # Init underlying normal distribution
-        self._normal = Normal(self.mean, self.std, is_symmetric=is_symmetric, **kwargs)
+        self._normal = Normal(self.mean, self.std, is_symmetric=True, **kwargs)
+
+    @property
+    def mean(self):
+        """ Mean of the distribution """
+        return self._mean
+
+    @mean.setter
+    def mean(self, value):
+        self._mean = force_ndarray(value, flatten=True)
+        if hasattr(self, '_normal'):
+            self._normal.mean = self._mean
+
+    @property
+    def std(self):
+        """ Std of the distribution """
+        return self._std
+
+    @std.setter
+    def std(self, value):
+        self._std = force_ndarray(value, flatten=True)
+        if hasattr(self, '_normal'):
+            self._normal.std = self._std
+
```
I wonder if we can simplify the logic a bit so we do not need to store local `_mean` and `_std`. I mean if we can set the initializer, and mean and std setters and getters to be something like this:

```python
    def __init__(self, mean=None, std=None, low=-np.Inf, high=np.Inf, is_symmetric=False, **kwargs):
        # Init from abstract distribution class
        super().__init__(is_symmetric=is_symmetric, **kwargs) 
        # Init specific to this distribution
        self.low = low
        self.high = high
        
        self._normal = Normal(mean, std, is_symmetric=True, **kwargs)

    @property
    def mean(self):
        """ Mean of the distribution """
        return self._normal.mean

    @mean.setter
    def mean(self, value):
        self._normal.mean = force_ndarray(value, flatten=True)

    @property
    def std(self):
        """ Std of the distribution """
        return self._normal.std

    @std.setter
    def std(self, value):
        self._normal.std = force_ndarray(value, flatten=True)

```

chaozg replied:
This simplified code seems to cause errors in initializing the Gibbs sampler, so after talking with @amal-ghamdi, we decided to keep it as is.

amal-ghamdi replied:
Maybe we can then just keep it as it is, unless @nabriis has other suggestions

nabriis replied:
No I think we cannot do it much simpler

amal-ghamdi commented on demos/demo38_truncated_normal.py:
```diff
@@ -0,0 +1,77 @@
+plt.ylim(-5, 5)
+plt.gca().set_aspect('equal')
+plt.title("Samples")
+# %% Demo 2
```
```suggestion

# %% Demo 2
```

amal-ghamdi reviewed (commented):
Hi @chaozg , many thanks for the update and the demo! 

I made a few small comments feel free to address them as you see fit.

nabriis reviewed (approved):
Thanks @chaozg! Nice work. LGTM

chaozg commented:
Thanks, @amal-ghamdi and @nabriis for your review! I'm merging to the main now
